src/OpenTK.Mathematics/Matrix/Matrix4.cs

```diff
@@ -1500,125 +1503,284 @@ public static void Mult(in Matrix4 left, float right, out Matrix4 result)
-                                {
-                                    maxPivot = absVal;
-                                    irow = j;
-                                    icol = k;
-                                }
-                            }
-                            else if (pivotIdx[k] > 0)
-                            {
-                                result = mat;
-                                return;
-                            }
-                        }
-                    }
-                }
-
-                ++pivotIdx[icol];
-
-                // Swap rows over so pivot is on diagonal
-                if (irow != icol)
-                {
-                    for (var k = 0; k < 4; ++k)
-                    {
-                        var f = inverse[irow, k];
-                        inverse[irow, k] = inverse[icol, k];
-                        inverse[icol, k] = f;
-                    }
-                }
-
-                rowIdx[i] = irow;
-                colIdx[i] = icol;
-
-                var pivot = inverse[icol, icol];
-
-                // check for singular matrix
-                if (pivot == 0.0f)
-                {
-                    throw new InvalidOperationException("Matrix is singular and cannot be inverted.");
-                }
-
-                // Scale row so it has a unit diagonal
-                var oneOverPivot = 1.0f / pivot;
-                inverse[icol, icol] = 1.0f;
-                for (var k = 0; k < 4; ++k)
-                {
-                    inverse[icol, k] *= oneOverPivot;
-                }
+                InvertFallback(in mat, out result);
+            }
+        }
 
-                // Do elimination of non-diagonal elements
-                for (var j = 0; j < 4; ++j)
-                {
-                    // check this isn't on the diagonal
-                    if (icol != j)
-                    {
-                        var f = inverse[j, icol];
-                        inverse[j, icol] = 0.0f;
-                        for (var k = 0; k < 4; ++k)
-                        {
-                            inverse[j, k] -= inverse[icol, k] * f;
-                        }
-                    }
-                }
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe void InvertSse3(in Matrix4 mat, out Matrix4 result)
+        {
+#pragma warning disable SA1114 // Parameter list should follow declaration
+#pragma warning disable SA1312 // Variable names should begin with lower-case letter
+#pragma warning disable SA1512 // Single-line comments should not be followed by blank line
+#pragma warning disable SA1515 // Single-line comment should be preceded by blank line
+
+            // Original derivation and implementation can be found here:
+            // https://lxjk.github.io/2017/09/03/Fast-4x4-Matrix-Inverse-with-SSE-SIMD-Explained.html
+
+            float* m = (float*)Unsafe.AsPointer(ref Unsafe.AsRef(in mat));
```

Comment:
You're not pinning here which could cause a GC bug, I believe.

You'll either have to pin or load the `Vector128<float>`s with `Unsafe.As`. I don't know which is faster.

Reply:
Yeah this is a mischievous bug indeed.
I'm my own test code I had already changed the code so that it was pinned.
Basically the only change to the assembly when using `fixed` is an extra move where the pointer is moved to the stack instead of being kept in registers.